### PR TITLE
Update NuGet v3 implementation to not show unlisted versions

### PR DESCRIPTION
**Pull request by maartenba**
This updates the NuGet v3 implementation to not show unlisted package versions (as described in https://github.com/NuGet/NuGetGallery/issues/3167)

### CODE REVIEW

**Comment by vip32** [comment]
thanks for making a fix for this annoying issue. please see https://github.com/badges/shields/issues/795

**Comment by ThadHouse** [comment]
Any update on getting this merged? It's definitely an issue with repos showing newer unlisted versions, especially since nuget doesn't allow removing packages.

File: server.js [hunk 1]
```diff
@@ -3492,27 +3492,49 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
+		var autocompleteResources = data.resources.filter(function(resource) {
+          return resource['@type'] == 'SearchAutocompleteService';
+        });
+		
+		// query autocomplete service
+		var reqUrl = autocompleteResources[0]['@id']
+          + '?id=' + id.toLowerCase()
+		  + '&prerelease=' + includePre
+		  + '&skip=0'
+		  + '&take=10000';
```
Comment (espadrine) [inline server.js]: Could you place comments explaining the purpose of each parameter please?

File: server.js [hunk 2]
```diff
@@ -3492,27 +3492,49 @@ function mapNugetFeed(pattern, offset, getInfo) {
   var vPreRegex = new RegExp('^\\/' + pattern + '\\/vpre\\/(.*)\\.(svg|png|gif|jpg|json)$');
 
   function getNugetVersion(apiUrl, id, includePre, request, done) {
-    var reqUrl = apiUrl + '/flatcontainer/' + id.toLowerCase() + '/index.json';
-    request(reqUrl, function(err, res, buffer) {
+	// get service index document
+	var serviceIndexUrl = apiUrl + '/index.json';	
+	request(serviceIndexUrl, function(err, res, buffer) {
```
Comment (espadrine) [inline server.js]: Having two successive requests will make this badge twice as slow to load. This request is superfluous.

Please directly call `https://api-v2v3search-0.nuget.org/autocomplete` (you are very welcome to alternate `-0-` and `-1-` to help their server load).

(Unless your assessment is that this endpoint changes faster than I can review patches, that is, about twice a year. In that case, please perform the request only once, and cache the result. For that, you can use the `regularUpdate()` function defined in server.js, with an interval corresponding to your estimated frequency of updates to the endpoint path.)

File: server.js (hunk 1 above)
Comment (maartenba) [inline server.js]: Fixed.

File: server.js [hunk 3]
```diff
@@ -3492,27 +3492,49 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
-        var versions = data.versions;
-        if (!includePre) {
-          // Remove prerelease versions.
-          filteredVersions = versions.filter(function(version) {
-            return !/-/.test(version);
-          });
-          if (filteredVersions.length > 0) {
-            versions = filteredVersions;
+		var autocompleteResources = data.resources.filter(function(resource) {
+          return resource['@type'] == 'SearchAutocompleteService';
```
Comment (maartenba) [inline server.js]: Fixed.

File: server.js (hunk 2 above)
Comment (maartenba) [inline server.js]: Fixed.

File: server.js [hunk 4]
```diff
@@ -3492,27 +3492,49 @@ function mapNugetFeed(pattern, offset, getInfo) {
   var vPreRegex = new RegExp('^\\/' + pattern + '\\/vpre\\/(.*)\\.(svg|png|gif|jpg|json)$');
 
   function getNugetVersion(apiUrl, id, includePre, request, done) {
-    var reqUrl = apiUrl + '/flatcontainer/' + id.toLowerCase() + '/index.json';
-    request(reqUrl, function(err, res, buffer) {
+	// get service index document
```
Comment (maartenba) [inline server.js]: Fixed.

File: server.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-var secureServer = !!process.env.HTTPS;
+﻿var secureServer = !!process.env.HTTPS;
```
Comment (maartenba) [inline server.js]: Should be good now. (all \r\n in the document)

**Comment by maartenba** [comment]
Processed comments.

/cc @skofman1 @joelverhagen

File: server.js (hunk 5 above)
Comment (espadrine) [inline server.js]: That is because the line is indeed different. You saved the file as UTF-8 with a byte order mark (U+FEFF).

File: server.js [hunk 6]
```diff
@@ -3492,29 +3492,52 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
-    request(reqUrl, function(err, res, buffer) {
-      if (err != null) {
-        done(err);
-        return;
-      }
-
-      try {
+    // get service index document
+    regularUpdate(apiUrl + '/index.json',
+      (3600 * 1000),
```
Comment (maartenba) [inline server.js]: It changes once a year, but they have the liberty to change it every 2 minutes. What would be a good value then?

File: server.js [hunk 7]
```diff
@@ -3492,29 +3492,52 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
+        return autocompleteResources;
+      },
+      function(err, autocompleteResources) {
+        if (err != null) { done(err); return; }
+
+        // query autocomplete service
+        var randomEndpointIdx = Math.floor(Math.random() * autocompleteResources.length);
+        var reqUrl = autocompleteResources[randomEndpointIdx]['@id']
+          + '?id=' + id.toLowerCase()    // NuGet package id (lowercase)
+          + '&prerelease=' + includePre  // Include prerelease versions?
```
Comment (maartenba) [inline server.js]: Done

File: server.js [hunk 8]
```diff
@@ -3492,29 +3492,52 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
+          + '&skip=0'                    // Start at first package found
+          + '&take=5000';                // Max. number of results
```
Comment (maartenba) [inline server.js]: It's the number of versions, so in essence we now limit this to packages that have 5000 packages max.

File: server.js (hunk 6 above)
Comment (espadrine) [inline server.js]: If it changes once a year, the average waiting time before the change is 6 months.

If you want, you can be fancy and say that the median waiting time before the endpoint changes is ln(2)/λ ≈ 0.69 years, assuming endpoint change is a Poisson process. But I don't know how independent the endpoint changes really are from each other.

But maybe you focused on minimizing the users' waiting time before getting correct information? In that case, assuming each request to fetch the endpoint makes users wait `l = 200 ms` for their badge, taking `x` to be the endpoint update period, the waiting time within a year is `l/x + x` in years, for which a derivation yields an optimum at `x = √l = √(200/1000/3600/24/365)` years, or roughly 41.856899073 minutes, which is pretty close to what you picked but not quite.

Whatever you pick, I'd rather it not be a magic number, and I'd rather the choice be explained and rationalized in a comment.

Pragmatically, if the endpoint changes, I can easily restart the servers to force them to get a new value.

Comment (espadrine) [inline server.js]: Another way to think about it is to tolerate a percentage `x` of all NuGet badge requests to fail. The probability of a single request to fail is `(update period) / (endpoint change period)`; the expected value of the number of failing requests is `(number of requests per year) * (update period) / (endpoint change period)` (expectation of a binomial distribution). So the update period is `(expected value of number of failing requests per year) * (endpoint change period) / (number of requests per year)`, ie. `(proportion of failing requests) * (endpoint change period)`. If we take the proportion of failing requests to be `10^-4` (0.01%), it comes out to 52.56 minutes.

File: server.js [hunk 9]
```diff
@@ -3492,29 +3492,52 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
+        return autocompleteResources;
+      },
+      function(err, autocompleteResources) {
+        if (err != null) { done(err); return; }
+
+        // query autocomplete service
+        var randomEndpointIdx = Math.floor(Math.random() * autocompleteResources.length);
+        var reqUrl = autocompleteResources[randomEndpointIdx]['@id']
+          + '?id=' + encodeURIComponent(id.toLowerCase())   // NuGet package id (lowercase)
+          + '&prerelease=' + encodeURIComponentincludePre)  // Include prerelease versions?
```
Comment (espadrine) [inline server.js]: Looks like there is a bit of a typo here.

**Comment by espadrine** [comment]
Could you look into why the `/dotnet.myget/dotnet-coreclr/v/Microsoft.DotNet.CoreCLR.svg` badge no longer works with this patch, please?

File: server.js (hunk 6 above)
Comment (maartenba) [inline server.js]: So... what value do we decide on? My take would be "1 day" then, given the NuGet team will do working days and probably only change once every working day (and like you say, a restart can always fix stale cache). Thoughts?

File: server.js (hunk 9 above)
Comment (maartenba) [inline server.js]: Ugh! Fixed :-) Thanks for spotting this.

**Comment by maartenba** [comment]
Fixed that badge.

**Comment by espadrine** [comment]
Thank you! Merged as 361a6e344a52aa06765c0c0336e65920a4ceeed7.

**Pull request closed without merging**